src/main/java/org/matsim/project/prebookingStudy/jsprit/PreplannedSchedulesCalculator.java

```diff
@@ -282,23 +283,27 @@ public double getCosts(VehicleRoutingProblemSolution solution) {
 			double numVehiclesUsed = solution.getRoutes().size();
 			double costForFleet = numVehiclesUsed * costPerVehicle;
 			if (!options.infiniteFleet) {
-				costForFleet = 0;
+				costForFleet = 0; // (when finite fleet is used, we can use all the vehicle as they are already there)
 			}
 
 			VehicleRoutingTransportCosts costMatrix = problem.getTransportCosts();
-			double totalTransportCost = 0;
-			for (VehicleRoute route : solution.getRoutes()) {
-				TourActivity prevAct = route.getStart();
-				for (TourActivity activity : route.getActivities()) {
-					totalTransportCost += costMatrix.getTransportCost(prevAct.getLocation(), activity.getLocation(),
-							prevAct.getEndTime(), route.getDriver(), route.getVehicle());
-					prevAct = activity;
-				}
-			}
-
-			totalTransportCost = totalTransportCost / 3600
-					* drivingCostPerHour;  // In current setup, transport cost = driving time
-			double totalCost = costForUnassignedRequests + costForFleet + totalTransportCost;
+            double totalTransportCost = 0;
+            double totalArrivalTime = 0; // The absolute value doesn't matter. We just want to minimize the sum arrival time (i.e., arrive early)
+            for (VehicleRoute route : solution.getRoutes()) {
+                TourActivity prevAct = route.getStart();
+                for (TourActivity activity : route.getActivities()) {
+                    if (activity.getName().equals("deliverShipment")) {
+                        totalArrivalTime += activity.getArrTime();
+                    }
+                    totalTransportCost += costMatrix.getTransportCost(prevAct.getLocation(), activity.getLocation(),
+                            prevAct.getEndTime(), route.getDriver(), route.getVehicle());
+                    prevAct = activity;
+                }
+            }
+
+            totalTransportCost = totalTransportCost / 3600 * drivingCostPerHour;  // In current setup, transport cost = driving time
+            double totalArrivalTimeCost = totalArrivalTime / 3600 * arrivalTimeCostPerHour;
+			double totalCost = costForUnassignedRequests + costForFleet + totalTransportCost + totalArrivalTimeCost;
```

Comment:
Unforunately, cost calcuation is used in jsprit at the end of each iteration. So during the iteration, there is no guarantee that individual insertions will be aligned with the solution cost calculator. The only way to ensure it is by changing the insertion cost calculation in jsprit, but that is a lot of work.

Reply:
I think we can still use the objective function to drive the system to where we want. I think if we provide the system with adequate iterations (e.g., now I have raised it from 1000 to 2000), then it will find the solution that corresponds to what we want.  

I guess it's fine that the ruin and recreate process just modify the plans without realizing our goal. But the threshold/upper limit of the costs will act as the selector. This is somewhat similar to the MATSim replanning. Agents just create random modifications to the original plans, and we will decide if we keep the plan based on the scoring function.

The advantage of keeping the same travel time constraint is that the solutions will be the same as before (when the same seed + single thread is used. Or we may even consider exporting the preplanned schedule). Then we can just determine if the solutions can withstand certain amount of fluctuation in traffic. The disadvantage for this approach is that it doesn't apply to DRT benchmark approach (and therefore, a change in max travel time or travel time estimation is required). 

Changing the travel time allowance is also a good option. And that enables a fairer comparison between the both approaches. The disadvantage is the needs for extra parameter (i.e., how much travel time are we going to reduce). 

We can discuss this in the meeting next week.